### PR TITLE
Fixed Sym get all projects

### PR DESCRIPTION
**Pull request by peterghrong**
# Summary

1. Fixed Sym's get_projects endpoint.
2. Added unittest for get_projects in test_projects.py
3. My requirements.txt did not pip correctly however, this is due to the psypopg2 package, I replaced it will psypopg2-binary and it was built correctly. 

	
## Test Plan

Added and Passed a quick unittest, added the test script in test_projects.py
Screen Shot 2021-01-14 at 11 42 55 AM

## Related Issues

Which issues does this PR resolve/work on?
Closes #66

### CODE REVIEW

File: backend/tests/test_project.py
Comment (CurtisMIT) [inline backend/tests/test_project.py:4]: @peterghrong I think this is an unused import?

**Comment by CurtisMIT** [comment]
@peterghrong You mentioned that you changed to `psycopg2-binary`, can you reflect the changes in your `requirements.txt`? A simple `pip freeze > requirements.txt` should do it.

File: backend/tests/test_project.py
Comment (CurtisMIT) [inline backend/tests/test_project.py:76]: @peterghrong Is it possible for you to change `test_project_routes` (a few lines above, Github won't let me comment there for some reason) to `test_update_project`? As now the naming would be a bit misleading with the new changes, which is a great idea to separate them so that we know what's failing and what's not instead of bundling all of them together in a massive fail. TYSM, should have foresaw that.

File: backend/tests/test_project.py
Comment (peterghrong) [reply backend/tests/test_project.py:4]: Yes, missed it mb

**Comment by peterghrong** [comment]
> @peterghrong You mentioned that you changed to `psycopg2-binary`, can you reflect the changes in your `requirements.txt`? A simple `pip freeze > requirements.txt` should do it.

I did it initially then it started failing the Travis CI build test on Github as you can see below.

File: backend/tests/test_project.py
Comment (peterghrong) [reply backend/tests/test_project.py:76]: fixed

File: backend/tests/test_project.py
Comment (CurtisMIT) [inline backend/tests/test_project.py:23]: TYSM

File: requirements.txt
Comment (CurtisMIT) [inline requirements.txt:1]: @peterghrong I think this file should be in the backend folder instead of the root directory.

File: backend/tests/test_project.py
Comment (CurtisMIT) [inline backend/tests/test_project.py:4]: @peterghrong I think you missed this?

**Review by CurtisMIT — COMMENTED** [review]
@peterghrong Yea don't worry about travis atm (not your fault), I saw a recurring issue with travis on that, I created a fix for it on #97 . Awaiting approval from celine.

File: backend/tests/test_project.py
Comment (peterghrong) [reply backend/tests/test_project.py:4]: Yep TYSM

File: requirements.txt
Comment (peterghrong) [reply requirements.txt]: You right legend, fixed !

File: backend/app/project/routes.py
Comment (CurtisMIT) [inline backend/app/project/routes.py:1]: @peterghrong unused import Flask.

File: backend/tests/test_project.py
Comment (CurtisMIT) [inline backend/tests/test_project.py:4]: @peterghrong Is there something wrong with your editor? This keeps on showing up (unused import request).

File: backend/app/project/routes.py
Comment (peterghrong) [reply backend/app/project/routes.py]: Removed, uuid is unused either

File: backend/tests/test_project.py
Comment (peterghrong) [reply backend/tests/test_project.py]: Removed

**Review by CurtisMIT — COMMENTED** [review]
@peterghrong LGTM once conflicts has been solved!

**Comment by celinehuang** [comment]
can I close #99 ?

**Comment by peterghrong** [comment]
Merge conflicts resolved and passed all tests

**Review by celinehuang — APPROVED** [review]
Looks very clean 🧼

File: backend/app/project/routes.py
Comment (CurtisMIT) [inline backend/app/project/routes.py:8]: @peterghrong minor issue, do we not want to leave it as just ""?

Comment (celinehuang) [reply backend/app/project/routes.py:8]: ^^

Comment (celinehuang) [reply backend/app/project/routes.py:8]: thanks @CurtisMIT

Comment (celinehuang) [reply backend/app/project/routes.py:8]: wait but also this is the same end point as here - https://github.com/hack4impact-mcgill/mu-crm-tool/blob/90cc6a54cb735e8ce85a0d428aea9e1386630530/backend/app/project/routes.py#L8 ??

Comment (peterghrong) [reply backend/app/project/routes.py:8]: I had ` @project.route("", methods=["GET"]) ` initially but now since "get_projects_of_type():" exists which it goes through the same route `@project.route("", methods=["GET"])`, it was not returning the right things. 

I also thought that it would be preferable to keep the get_all routes consistent.
eg. `@project.route("/types", methods=["GET"])`

Regarding the other pr, I pulled this branch into the other branch for some reason, it was an amateur mistake and I should have just undid the git pull before I opened that pr. I'm really sorry for all this confusion. 

Lemme know how you still think if it's better to  change it to `@project.route("", methods=["GET"]) `, I'll get it fixed ASAP

Comment (CurtisMIT) [reply backend/app/project/routes.py:8]: @peterghrong @celinehuang Yea in hindsight I maybe should have seen this coming, thanks for pointing out the clash again, mb missed it again. When we're sorting/filtering for more than one parameter, would we want to do it in the style below instead? As idk if we want to create a separate route for every filter we want to apply for projects. I've rewritten it so that we can freely add future request arguments if needed. It compiles and passes the tests. Any thoughts?

```
# get projects according to specified arguments (if any)
@project.route("", methods=["GET"])
def get_projects():
    projects = Project.query.all()

    type = request.args.get("type")
    if type is not None:
        if type == "":
            abort(404, "Project type invalid")
        else:
            projects = list(filter(lambda project: (project.type == type), projects))
    
    # add addition request arguments below

    return jsonify(Project.serialize_list(projects))
```

Inspired by these links:

scroll to (allow filtering, sorting, and pagination)
- https://stackoverflow.blog/2020/03/02/best-practices-for-rest-api-design/

scroll to (Using Query Arguments)
- https://www.digitalocean.com/community/tutorials/processing-incoming-request-data-in-flask

Comment (celinehuang) [reply backend/app/project/routes.py:8]: hmm it's actually good to have the return all project type 'helper' endpoint because in frontend we have to render all possible filters anyways? Although now that I look at it, should we add a query parameter in the endpoint..?

Comment (CurtisMIT) [reply backend/app/project/routes.py:8]: Yea I agree with the `"/types"` endpoint. The code segment I pasted is addressing the `""` endpoint, so that any filtering of projects can be done through that end point. To resolve the conflict for the `""` route.

Comment (celinehuang) [reply backend/app/project/routes.py:8]: Thanks Curtis for bringing up the more proper conventions we should follow when creating endpoints. @peterghrong the snippet Curtis posted is a lot cleaner than the way I had originally suggested so I think we can remove this endpoint https://github.com/hack4impact-mcgill/mu-crm-tool/blob/28bdaf8681c291d7e46fb1ea01216bb2c64716f3/backend/app/project/routes.py#L16  that's in main and replace it with what Curtis had suggested

**Review by celinehuang — APPROVED** [review]
LGTM 💯

**Pull request merged**
